app/queries/keyset_pagination.rb

```diff
@@ -0,0 +1,130 @@
+module Queries
+  class KeysetPagination
+    attr_reader :client, :order, :key, :count, :previous
+
+    # Initialises the keyset pagination class.
+    # Params:
+    # +client+:: the pagination client to query over
+    # +key+:: +Hash+ a hash containing the pagination key, mapped from presented name to internal name, i.e. { id: "editions.id" }
+    # +order+:: +Symbol+ either :asc for ascending and :desc for descending
+    # +count+:: +Fixnum+ the number of records to return in each page
+    # +after+:: +Array+ the current page to paginate after, an array containing a value for each field in the key
+    def initialize(client, key: nil, order: nil, count: nil, before:, after:)
+      @client = client
+      @key = key || { id: "id" }
+      @order = order || :asc
+      @count = (count || 100).to_i
+
+      if before.present? && after.present?
+        raise "Before and after cannot both be present."
+      end
+
+      if before
+        @previous = before
+        @order = order == :asc ? :desc : :asc
+        @should_reverse = true
+      else
+        @previous = after
+        @should_reverse = false
+      end
+
+      if previous.present? && previous.count != key.count
+        raise "Number of previous values does not match the number of fields."
+      end
+    end
+
+    def call
+      results
+    end
+
+    def next_before_key
+      key_for_record(results.first)
+    end
+
+    def next_after_key
+      key_for_record(results.last)
+    end
+
+    def has_next_before?
+      KeysetPagination.new(
+        client, key: key, order: order, count: 1,
+        before: next_before_key, after: nil
+      ).call.count >= 1
```

Comment:
I'm not sure that it is overkill.

There would be a lot of duplication when generating the `where_next_page` query like that. It wouldn't be possible to use `paginated_query` as it is right now on both because one goes forwards and one goes backwards, so all the logic which reverses the query would also need to be passed into that method rather than in the constructor of the `KeysetPagination` class -- in such a way that the `paginated_query` method ends up becoming as complex as the `KeysetPagination` class itself.